aioxmpp/entitycaps/service.py

```diff
@@ -334,50 +428,86 @@ def lookup_info(self, jid, keys):
 
     @aioxmpp.service.outbound_presence_filter
     def handle_outbound_presence(self, presence):
-        if (presence.type_ == aioxmpp.structs.PresenceType.AVAILABLE and
-                self.__current_keys):
-            self.logger.debug("injecting capabilities into outbound presence")
-            self.__115.put_keys(self.__current_keys, presence)
+        if (presence.type_ == aioxmpp.structs.PresenceType.AVAILABLE
+                and self.__active_hashsets):
+            current_hashset = self.__active_hashsets[-1]
+
+            try:
+                keys = current_hashset[self.__115]
+            except KeyError:
+                pass
+            else:
+                self.__115.put_keys(keys, presence)
+
+            try:
+                keys = current_hashset[self.__390]
+            except KeyError:
+                pass
+            else:
+                self.__390.put_keys(keys, presence)
 
         return presence
 
     @aioxmpp.service.inbound_presence_filter
     def handle_inbound_presence(self, presence):
-        keys = list(self.__115.extract_keys(presence))
-        lookup_task = asyncio.async(
-            self.lookup_info(presence.from_, keys)
-        )
-        self.disco_client.set_info_future(presence.from_, None, lookup_task)
+        keys = []
+
+        if self.xep390_support:
+            keys.extend(self.__390.extract_keys(presence))
+
+        if self.xep115_support:
+            keys.extend(self.__115.extract_keys(presence))
+
+        if keys:
+            lookup_task = asyncio.async(
```

Comment:
> […] but this is the standard library we modify […]

This is exactly why I refrain from it.

> only a specific set of versions will be affected (since all later versions have ``ensure_future`), on the other hand I have to admit, that we'll discontinue Python 3.4 support once Python versions allowing ``async def`` will be widespread, which hopefully will be long before a Python version reserving ``async`` as a keyword will be widespread.

We have a lot places throughout the code where ``asyncio.async`` is used. This is a matter for #11. As long as we don’t support a version of python which makes ``async`` a keyword globally, I don’t think we should rush to a change here. Once upstream makes the move, we’ll reconsider. In the end, it’ll approximately be a simple ``sed`` over the tree to fix things.

> Another workaround could be to use a aioxmpp.compat module (and define ensure_future there, depending on what is available, Plus, we will probably need a compatibility layer at some point anyways).

I would like to avoid that as far as sensible.